File: server/ingest-historical.mjs
/**
 * Historical Fireflies export ingest pipeline.
 *
 * Usage (CLI):
 *   node server/ingest-historical.mjs <path-to-historical_data-dir>
 *
 * Programmatic:
 *   import { fireflies_historical } from "./ingest-historical.mjs";
 *   await fireflies_historical("/abs/path/to/historical_data");
 */

import fs from "fs";
import path from "path";
import { upsertMeetings } from "./sqlite-api.mjs";
import { buildMeetingMetadata, classifyMeetingType } from "./apis.mjs";

// ─────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────

function readJsonFile(filePath) {
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch {
    return null;
  }
}

function readTextFile(filePath) {
  try {
    return fs.readFileSync(filePath, "utf8").trim();
  } catch {
    return null;
  }
}

/**
 * Given a meeting folder, find a file matching a glob-style prefix pattern
 * and return its full path + the extracted ID segment.
 *
 * e.g. prefix="meeting-metadata-", suffix=".txt"
 * matches "meeting-metadata-7wE5Rlay1au0TttC.txt" → id="7wE5Rlay1au0TttC"
 */
function findFileByPattern(dir, prefix, suffix) {
  const entries = fs.readdirSync(dir);
  for (const entry of entries) {
    if (entry.startsWith(prefix) && entry.endsWith(suffix)) {
      const id = entry.slice(prefix.length, -suffix.length);
      return { filePath: path.join(dir, entry), id };
    }
  }
  return null;
}

/**
 * Resolve the organizer / host email: prefer the first itilite.com address,
 * otherwise fall back to the first attendee.
 */
function resolveOrganizerEmail(attendees) {
  if (!Array.isArray(attendees) || attendees.length === 0) return null;
  const internal = attendees.find((e) => e && e.toLowerCase().includes("@itilite.com"));
  return internal || attendees[0] || null;
}

// ─────────────────────────────────────────────
// Per-meeting parser
// ─────────────────────────────────────────────

function parseMeetingFolder(meetingDir) {
  // 1. Locate the metadata file — its suffix gives us the meeting ID
  const metaMatch = findFileByPattern(meetingDir, "meeting-metadata-", ".txt");
  if (!metaMatch) {
    console.warn(`  [skip] No meeting-metadata file in ${path.basename(meetingDir)}`);
    return null;
  }

  const { filePath: metaPath, id } = metaMatch;

  const rawMeta = readJsonFile(metaPath);
  if (!rawMeta) {
    console.warn(`  [skip] Could not parse metadata in ${path.basename(meetingDir)}`);
    return null;
  }

  // 2. Read optional companion files
  const summaryMatch = findFileByPattern(meetingDir, "meeting-summary-", ".txt");
  const urlMatch = findFileByPattern(meetingDir, "meeting-url-", ".txt");

  const summaryText = summaryMatch ? readTextFile(summaryMatch.filePath) : null;
  const transcriptUrl = urlMatch ? readTextFile(urlMatch.filePath) : null;

  const speakerMeta = readJsonFile(path.join(meetingDir, "speaker-meta.json"));
  const transcriptData = readJsonFile(path.join(meetingDir, "transcript.json"));

  // 3. Build the normalised meeting object
  const attendees = Array.isArray(rawMeta.attendees) ? rawMeta.attendees.filter(Boolean) : [];
  const organizerEmail = resolveOrganizerEmail(attendees);

  const startMs = rawMeta.meetingStartTime
    ? new Date(rawMeta.meetingStartTime).getTime()
    : null;

  // duration from Fireflies export is in minutes; keep consistent with live sync
  const durationMinutes =
    typeof rawMeta.meetingDuration === "number" ? rawMeta.meetingDuration : null;

  const meeting = {
    id,
    title: rawMeta.meetingTitle || null,
    date: startMs,
    duration: durationMinutes,
    organizer_email: organizerEmail,
    host_email: organizerEmail,          // same source for historical data
    transcript_url: transcriptUrl || null,
    audio_url: null,                     // local files not served
    video_url: null,
    meeting_type: null,                  // filled below after classification
    participants: attendees,
    meeting_attendees: attendees.map((email) => ({ email })),
    user: null,
    summary: summaryText || null,        // raw AI summary block as a single string
    sentences: transcriptData?.data ?? [],
    metadata: null,                      // filled below after buildMeetingMetadata
    hubspot_deals: [],
    // store speaker mapping in metadata extension
    _speakerMeta: speakerMeta?.speakerMeta ?? null,
  };

  // 4. Derive meeting_type and metadata using the shared helpers
  meeting.meeting_type = classifyMeetingType(meeting);
  meeting.metadata = {
    ...buildMeetingMetadata(meeting),
    source: "historical_export",
    speakerMeta: meeting._speakerMeta,
    rawStartTime: rawMeta.meetingStartTime || null,
    rawEndTime: rawMeta.meetingEndTime || null,
  };
  delete meeting._speakerMeta;

  return meeting;
}

// ─────────────────────────────────────────────
// Main export
// ─────────────────────────────────────────────

/**
 * Process all meetings inside a Fireflies historical export directory and
 * upsert them into the SQLite `meetings` table.
 *
 * @param {string} historicalDir
 *   Absolute (or CWD-relative) path to the root historical_data folder.
 *   The function descends one level into export-* sub-folders automatically.
 *
 * @param {{ batchSize?: number, onProgress?: (done: number, total: number) => void }} [opts]
 *
 * @returns {Promise<{ inserted: number, skipped: number, errors: string[] }>}
 */
export async function fireflies_historical(historicalDir, opts = {}) {
  const { batchSize = 20, onProgress } = opts;

  const absRoot = path.resolve(historicalDir);
  if (!fs.existsSync(absRoot)) {
    throw new Error(`Historical data directory not found: ${absRoot}`);
  }

  // Collect all meeting-level directories (two levels deep: root → export-* → meeting-*)
  const meetingDirs = [];
  const topEntries = fs.readdirSync(absRoot, { withFileTypes: true });

  for (const exportEntry of topEntries) {
    if (!exportEntry.isDirectory()) continue;
    const exportDir = path.join(absRoot, exportEntry.name);
    const meetingEntries = fs.readdirSync(exportDir, { withFileTypes: true });
    for (const meetingEntry of meetingEntries) {
      if (meetingEntry.isDirectory()) {
        meetingDirs.push(path.join(exportDir, meetingEntry.name));
      }
    }
  }

  console.log(`[historical] Found ${meetingDirs.length} meeting folder(s) in ${absRoot}`);

  let inserted = 0;
  let skipped = 0;
  const errors = [];
  const batch = [];

  const flush = () => {
    if (batch.length === 0) return;
    try {
      upsertMeetings(batch.splice(0));
    } catch (err) {
      errors.push(`Batch upsert failed: ${err.message}`);
    }
  };

  for (let i = 0; i < meetingDirs.length; i++) {
    const dir = meetingDirs[i];
    const label = path.basename(dir);

    try {
      const meeting = parseMeetingFolder(dir);
      if (!meeting) {
        skipped++;
      } else {
        batch.push(meeting);
        inserted++;
        console.log(`  [ok] ${label} → id=${meeting.id} type=${meeting.meeting_type}`);
      }
    } catch (err) {
      skipped++;
      errors.push(`${label}: ${err.message}`);
      console.error(`  [err] ${label}: ${err.message}`);
    }

    if (batch.length >= batchSize) flush();
    if (onProgress) onProgress(i + 1, meetingDirs.length);
  }

  flush(); // remainder

  console.log(
    `[historical] Done — inserted/updated: ${inserted}, skipped: ${skipped}, errors: ${errors.length}`,
  );

  return { inserted, skipped, errors };
}

// ─────────────────────────────────────────────
// CLI entry point
// ─────────────────────────────────────────────

if (process.argv[1] && path.resolve(process.argv[1]) === path.resolve(import.meta.url.replace("file:///", "").replace(/\//g, path.sep))) {
  const dir = process.argv[2];
  if (!dir) {
    console.error("Usage: node server/ingest-historical.mjs <path-to-historical_data>");
    process.exit(1);
  }

  fireflies_historical(dir, {
    onProgress: (done, total) => {
      if (done % 10 === 0 || done === total) {
        process.stdout.write(`\r  Progress: ${done}/${total}`);
      }
    },
  })
    .then(({ inserted, skipped, errors }) => {
      process.stdout.write("\n");
      if (errors.length) {
        console.error("Errors:");
        errors.forEach((e) => console.error("  •", e));
      }
      process.exit(errors.length > 0 ? 1 : 0);
    })
    .catch((err) => {
      console.error("Fatal:", err.message);
      process.exit(1);
    });
}

File: backend/utils/ingest_historical.py
"""
ingest_historical.py — Fireflies historical export ingest pipeline.
Mirrors ingest-historical.mjs.
"""

import json
import os
from pathlib import Path

from db.merlin_db import upsert_meetings
from apis.fireflies import classify_meeting_type, build_meeting_metadata


# ─── Helpers ──────────────────────────────────────────────────────────────────

def _read_json(file_path: str) -> dict | None:
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return None


def _read_text(file_path: str) -> str | None:
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return f.read().strip()
    except Exception:
        return None


def _find_file_by_pattern(directory: str, prefix: str, suffix: str) -> dict | None:
    """Find a file matching prefix+id+suffix, return {'file_path': ..., 'id': ...}."""
    try:
        for entry in os.listdir(directory):
            if entry.startswith(prefix) and entry.endswith(suffix):
                meeting_id = entry[len(prefix):-len(suffix)]
                return {"file_path": os.path.join(directory, entry), "id": meeting_id}
    except Exception:
        pass
    return None


def _resolve_organizer_email(attendees: list[str]) -> str | None:
    if not attendees:
        return None
    internal = next((e for e in attendees if e and "@itilite.com" in e.lower()), None)
    return internal or attendees[0]


def _parse_meeting_folder(meeting_dir: str) -> dict | None:
    meta_match = _find_file_by_pattern(meeting_dir, "meeting-metadata-", ".txt")
    if not meta_match:
        print(f"  [skip] No meeting-metadata file in {os.path.basename(meeting_dir)}")
        return None

    raw_meta = _read_json(meta_match["file_path"])
    if not raw_meta:
        print(f"  [skip] Could not parse metadata in {os.path.basename(meeting_dir)}")
        return None

    meeting_id = meta_match["id"]

    summary_match = _find_file_by_pattern(meeting_dir, "meeting-summary-", ".txt")
    url_match = _find_file_by_pattern(meeting_dir, "meeting-url-", ".txt")

    summary_text = _read_text(summary_match["file_path"]) if summary_match else None
    transcript_url = _read_text(url_match["file_path"]) if url_match else None

    speaker_meta = _read_json(os.path.join(meeting_dir, "speaker-meta.json"))
    transcript_data = _read_json(os.path.join(meeting_dir, "transcript.json"))

    attendees = [a for a in (raw_meta.get("attendees") or []) if a]
    organizer_email = _resolve_organizer_email(attendees)

    start_ms = None
    if raw_meta.get("meetingStartTime"):
        from datetime import datetime
        try:
            start_ms = int(datetime.fromisoformat(raw_meta["meetingStartTime"].replace("Z", "+00:00")).timestamp() * 1000)
        except Exception:
            pass

    duration_minutes = raw_meta.get("meetingDuration") if isinstance(raw_meta.get("meetingDuration"), (int, float)) else None

    meeting: dict = {
        "id":              meeting_id,
        "title":           raw_meta.get("meetingTitle"),
        "date":            start_ms,
        "duration":        duration_minutes,
        "organizer_email": organizer_email,
        "host_email":      organizer_email,
        "transcript_url":  transcript_url,
        "audio_url":       None,
        "video_url":       None,
        "meeting_type":    None,
        "participants":    attendees,
        "meeting_attendees": [{"email": e} for e in attendees],
        "user":            None,
        "summary":         summary_text,
        "sentences":       (transcript_data.get("data") if transcript_data else None) or [],
        "metadata":        None,
        "hubspot_deals":   [],
    }

    meeting["meeting_type"] = classify_meeting_type(meeting)
    speaker_meta_val = (speaker_meta.get("speakerMeta") if speaker_meta else None)
    meeting["metadata"] = {
        **build_meeting_metadata(meeting),
        "source":       "historical_export",
        "speakerMeta":  speaker_meta_val,
        "rawStartTime": raw_meta.get("meetingStartTime"),
        "rawEndTime":   raw_meta.get("meetingEndTime"),
    }

    return meeting


# ─── Main entry ───────────────────────────────────────────────────────────────

async def fireflies_historical(
    historical_dir: str,
    *,
    batch_size: int = 20,
    on_progress=None,
) -> dict:
    abs_root = os.path.abspath(historical_dir)
    if not os.path.exists(abs_root):
        raise FileNotFoundError(f"Historical data directory not found: {abs_root}")

    meeting_dirs: list[str] = []
    for export_entry in os.scandir(abs_root):
        if not export_entry.is_dir():
            continue
        for meeting_entry in os.scandir(export_entry.path):
            if meeting_entry.is_dir():
                meeting_dirs.append(meeting_entry.path)

    print(f"[historical] Found {len(meeting_dirs)} meeting folder(s) in {abs_root}")

    inserted = 0
    skipped  = 0
    errors:  list[str] = []
    batch:   list[dict] = []

    def flush():
        nonlocal errors
        if not batch:
            return
        try:
            upsert_meetings(batch[:])
            batch.clear()
        except Exception as err:
            errors.append(f"Batch upsert failed: {err}")
            batch.clear()

    for i, d in enumerate(meeting_dirs):
        label = os.path.basename(d)
        try:
            meeting = _parse_meeting_folder(d)
            if not meeting:
                skipped += 1
            else:
                batch.append(meeting)
                inserted += 1
                print(f"  [ok] {label} → id={meeting['id']} type={meeting['meeting_type']}")
        except Exception as err:
            skipped += 1
            errors.append(f"{label}: {err}")
            print(f"  [err] {label}: {err}")

        if len(batch) >= batch_size:
            flush()
        if on_progress:
            on_progress(i + 1, len(meeting_dirs))

    flush()

    print(f"[historical] Done — inserted/updated: {inserted}, skipped: {skipped}, errors: {len(errors)}")
    return {"inserted": inserted, "skipped": skipped, "errors": errors}
